### PR TITLE
compose: Disable the topic input on `emptyTopicOnly` channels.

### LINKED ISSUE #1843: Hide topic field on channels restricted to "general chat" only
On the web version, the topic box is hidden on channels where topics are disabled (i.e. general chat only). On this mobile client, the topic field is available, and setting the value results in an error.

I think it's much more intuitive to hide the topic box on mobile in channels that are general chat only.

### PR DESCRIPTION
**Pull request by Ruhdee**
Fixes #1843.

Screenshots/Screen captures

|Before|After|
|----|----|
|before|after|

| Before | After |
| --- | --- |
| before | after |

This PR implements the widget changes associated with `emptyTopicOnly` policy for channels, which was introduced in #2231. Although the issue description states that topic input needs to be hidden, as discussed in [#mobile-design > Changes for per-channel topics policy @ 💬](https://chat.zulip.org/#narrow/channel/530-mobile-design/topic/Changes.20for.20per-channel.20topics.20policy/near/2417010), disabling the topic input is the correct behavior.

This PR also moves the topics-policy resolution tests into test/model/channel_test.dart. Ideally, those tests should have been introduced as model tests in #2231 itself.

### CODE REVIEW

**Review by kenclary — APPROVED** [review]
This looks very good, especially with the tests moved. Maintainer review can triple check that no tests got lost in the shuffle.

**Review by rajveermalviya — APPROVED** [review]
Thanks @Ruhdee! All LGTM, moving over to Chris' review.

File: test/widgets/compose_box_test.dart [hunk 1]
```diff
@@ -688,6 +688,40 @@ void main() {
...
+          topicHintText: eg.defaultRealmEmptyTopicDisplayName,
+          contentHintText: 'Message #${channel.name} > '
+                           '${eg.defaultRealmEmptyTopicDisplayName}');
+        check(tester.widget<TextField>(topicInputFinder))
+          ..enabled.equals(false)
+          ..decoration.isNotNull().hintStyle.isNotNull()
+            .fontStyle.equals(FontStyle.italic);
+      });
+
+      testWidgets('focuses content, not topic input', (tester) async {
```
Comment (chrisbobbe) [inline test/widgets/compose_box_test.dart]: This test is really about focus, not hint text; how about moving it to the 'auto focus' group? Also see the existing "empty fetch" tests there for a natural way to trigger `requestFocusIfUnfocused`, vs. calling it directly.

**Review by chrisbobbe — COMMENTED** [review]
Thanks! This looks mostly good.

Claude flagged one bug, which is that you can quote-and-reply an old message in a non-empty topic (from before the policy changed to empty-topic only), and the topic input will prefill with that non-empty topic, and there'd be no way to clear it out. Can fix by making `ComposeTopicController.setTopic` a no-op when the policy is empty-topic-only, e.g.:

```dart
  /// Set the input's text to [newTopic].
  ///
  /// Does nothing when the topic isn't the user's to choose
  /// (see [emptyTopicOnly]).
  void setTopic(TopicName newTopic) {
    if (emptyTopicOnly) return;
    value = TextEditingValue(text: newTopic.displayName ?? '');
  }
```

I think I'd prefer that, vs. editing the quote-and-reply caller. With #668, the other `setTopic` callers would be vulnerable to a similar bug: selecting an autocomplete option, and restoring a message that failed to send.

**Comment by Ruhdee** [comment]
@chrisbobbe Thank you for the review! This issue turned out to be more complex than I anticipated!

* From manual testing, I've found that the `setTopic` bug isn't possible in the quote-and-reply case. A channel's policy can only be switched to emptyTopicsOnly when there is a single topic named 'general chat.' Otherwise, admins need to merge and rename topics before switching the policy. Thus, a message cannot exist outside the empty topic. However, restoring an outbox message did surface the bug you mentioned, so I have addressed it.
* Changes in `_topicOrContentFocusChanged` were needed to handle an edge case (an assert statement was throwing). 
* While looking into the `setTopic` bug, I found another bug: If there is some text in the topic input, and the policy gets live updated to empty topic only, then the text in the now disabled topic input becomes stuck, and sending the message gives an error banner. I added a listener on store to fix it in a separate commit (can be squashed if desired). The same listener can be used to wire the updates from #668 in the future.
* Added tests to cover all these.

PTAL

**Comment by chrisbobbe** [comment]
> * Changes in `_topicOrContentFocusChanged` were needed to handle an edge case (an assert statement was throwing).

I looked into this, and I think the `assert` is actually wrong and the cleanest fix is to remove it (rather than add stateful fields `_topicHadFocus`/`_contentHadFocus`). Would you take a look at #2366, which does that? If that looks good, we can plan to merge that first, as prep for this PR.

**Comment by Ruhdee** [comment]
@chrisbobbe Thank you for #2366, rebased onto main. PTAL!

**Comment by chrisbobbe** [comment]
Thanks!

It looks like the commit sequence in this revision—

44df0f992 channel test: Move topics-policy resolution tests to model.
2e463c57f compose: Disable the topic input on `emptyTopicOnly` channels.
189a364d2 compose test [nfc]: Extract event-queue-expiry helpers.
e5d1225b4 compose: Clear the topic input when a channel becomes `emptyTopicOnly`.

—includes a known-broken state, from the second commit ("Disable the topic input […]") until the fourth commit ("Clear the topic input […]"):

```
[…] Before this, the policy could change while there was text in
the topic input. The newly disabled input would give no way to
remove that text, and sending would use it as the topic, which the
server would reject.
```

Would you update the branch to avoid that? We should be able to reverse the order of the two behavior changes, I think. A practice where each commit is independently deployable simplifies reading Git history, makes it easier to revert changes as necessary, and makes `git bisect` more powerful by not requiring developers to think about known-broken states. 🙂

**Comment by Ruhdee** [comment]
@chrisbobbe Thank you for pointing that out. I have reordered the commits. PTAL.

**Comment by chrisbobbe** [comment]
Thanks, this looks great! Merging.
